File: Ariel/PPAriel.py
# pylint: disable=import-error
import numpy as np
import config as cfg
bodyname = 'Ariel'
bodycode = 701

import spiceypy as spice
spice.furnsh('Utilities/spice/'+cfg.spicePCK)
_, (a,b,c) = spice.bodvcd(bodycode, 'RADII', 3)
R = np.sqrt((a**2 + b**2 + c**2)/3)

### Custom changes made to these profiles

# Overwrites the default values using user-defined lists

# add "field : [profile 1 value, profile 2 value, ...]"
# make sure all lists have the same length
"""
Example 1:
Producing 3 profiles.
1st profile has (Tb_K,ocean_wpct) = (270.0 , 10)
2nd profile has (Tb_K,ocean_wpct) = (271.0 , 5)
3rd profile has (Tb_K,ocean_wpct) = (272.0 , 0)
##############
CustomDict = {
    'Tb_K' : [270.0 , 271.0 , 272.0],
    'ocean_wpct' : [10 , 5 , 0]
}

numProfiles = 3
###############

Example 2:
Using np.linspace and meshgrid to create profiles on a "grid".
We want 25 profiles, one for each pair of
    5 values of Tb_K: [270.0,270.5,271.0,271.5,272.0]
    5 values of ocean_wpct: [0 , 2 , 4 , 6 , 8]
###############
Tb_K_vals = np.linspace(270.0 , 272.0 , 5)
ocean_wpct_vals = np.linspace(0 , 8 , 5)
Tb_K_arr , ocean_wpct_arr = np.meshgrid(Tb_K_vals , ocean_wpct_vals)

CustomDict = {
    'Tb_K' : Tb_K_arr.flatten(),
    'ocean_wpct' : ocean_wpct_arr.flatten()
}

numProfiles = 25
###############
"""

Tb_K_vals = np.linspace(270.0 , 272.0 , 5)
ocean_wpct_vals = np.linspace(0 , 8 , 5)
Tb_K_arr , ocean_wpct_arr = np.meshgrid(Tb_K_vals , ocean_wpct_vals)

CustomDict = {
    'Tb_K' : Tb_K_arr.flatten(),
    'ocean_wpct' : ocean_wpct_arr.flatten()
}

numProfiles = 25

### Construction of default dictionary for Europa

# Orbital and plotting parameters for use in LayeredInduction Response
OrbitalDict = {
    'peaks_Hz' : [3.45599e-5, 2.30405e-5, 1.1519e-5],
    'peaks_Hr' : [1/3600/i for i in [3.45599e-5, 2.30405e-5, 1.1519e-5]],
    'f_orb' : 2*np.pi/2.520/86400, # frequency of Ariel's orbit [rad/s]
#    'ionos_bounds' : 100e3,
#    'ionosPedersen_sig' : 30/100e3,
    'ionos_only' : [],
    'PLOT_SIGS' : True,
    'ADD_TRANSITION_BOUNDS' : False
}

# Bulk and surface properties
# note: C/MR2 uses value from Anderson et al. 1998
BulkSurfaceDict = {
    'rho_kgm3' : 1592., # average density [kg/m^3]
    'R_m' : R*1e3, # radius [m]
    'M_kg' : 1.353e21, # mass [kg]
#    'gsurf_ms2' : 1.428,
    'Tsurf_K' : 60.0,
    'Psurf_MPa' : 0.0,
    'Cmeasured' : 0.306, # C/MR^2 value
    'Cuncertainty' : 0.08, # C/MR^2 uncertainty
    'FeCore' : False, # boolean for whether planet has an iron core
    'rhoFe' : 8000., # density of pure iron
    'rhoFeS' : 5150. # density of iron sulfate
}

# Mantle heat properties
# cold case
MantleHeatDict = {
    'kr_mantle' : 4., # rock conductivity, (Cammarano et al. 2006, Table 4)
    'Qmantle_Wm2' : 2.7e4 / 4 / np.pi / BulkSurfaceDict['R_m']**2,
    'QHmantle' : 0.,
    'EQUIL_Q' : False
}

# Porosity of the rock
PorosityDict = {
    'POROUS_ROCK' : True,
    'phi_surface' : 0.8
}

# Core properties
CoreDict = {
#    'xFeS_meteoritic' : 0.0405,
    'xFeS' : 0.25,
#    'xFe_core' : 0.0279,
#    'XH2O' : 0.0035,
    'rho_sil_withcore_kgm3' : 2700.
}

# Ocean properties (included as sub-dictionary with key 'Ocean')
OceanDict = {
    'ocean_comp' : 'NH3', # composition of the ocean
    'ocean_wpct' : 3., # % concentration of solute in ocean
    'Tb_K' : 270. # temperature at the bottom of the ice layer
}

# combines all the dictionaries into one containing all the fields
PlanetDict = {
    'name' : bodyname,
    **OrbitalDict,
    **BulkSurfaceDict,
    **MantleHeatDict,
    **PorosityDict,
    **CoreDict,
    **OceanDict
}

### converts a 'dictionary' to a structured array with 'num' repetitions of said dictionary (for multiple profiles)

def dictionaryToStructuredArray(dictionary , num):
    keys = list(dictionary.keys())

    dt = [] # use list instead of numpy since dtype cannot be an array

    for i in range(len(keys)):
        if type(dictionary[keys[i]]) == str:
            dt.append( (keys[i],f'U{len(dictionary[keys[i]])}')) # resolves issue with string datatype in structured array
        else:
            dt.append( (keys[i] , type(dictionary[keys[i]]) ) ) # generate the dtype list for the structured array

    output = np.empty( (num) , dtype=dt ) # preallocate structured array

    for i in range(len(keys)):
        output[0][keys[i]] = dictionary[keys[i]] # create the first profile

    for i in range(1,num):
        output[i] = output[0] # add remaining profiles

    return output

Planet = dictionaryToStructuredArray(PlanetDict , numProfiles)

### amends Planets structured array using custom changes made in first section

for key in CustomDict.keys():
    Planet[key] = CustomDict[key]


### Creating the seismic dictionary
# Attenuation parameters based on those described in Cammarano et al. 2006
IceI = {
    'B_aniso_iceI' : 0.56,
    'gamma_atten_iceI' : 0.2,
    'g_aniso_iceI' : 22.
}
IceII = {
    'B_aniso_iceII' : 0.56,
    'gamma_atten_iceII' : 0.2,
    'g_aniso_iceII' : 25.
}
IceIII = {
    'B_aniso_iceIII' : 0.56,
    'gamma_atten_iceIII' : 0.2,
    'g_aniso_iceIII' : 27.
}
IceV = {
    'B_aniso_iceV' : 0.56,
    'gamma_atten_iceV' : 0.2,
    'g_aniso_iceV' : 28.
}
IceVI = {
    'B_aniso_iceVI' : 0.56,
    'gamma_atten_iceVI' : 0.2,
    'g_aniso_iceVI' : 30.
}
Mantle = {
    'B_aniso_mantle' : 0.56,
    'gamma_atten_mantle' : 0.2,
    'g_aniso_mantle': 30.
}

Seismic = {
    'LOW_ICE_Q' : 1.,
    'QScore' : 1e4,
#    'coreEOS' : 'sulfur_core_partition_SE15_1pctSulfur.tab',
    'mantleEOS' : 'echon_678_1.tab',
    **IceI,
    **IceII,
    **IceIII,
    **IceV,
    **IceVI,
    **Mantle
}

### Creating the Params dictionary

Params = {
    'cfg' : cfg,
    'wlims' : [np.log(0.001),np.log(1000)],
    'foursubplots' : True,
    'Legend' : False,
    'LegendPosition' : 'North',
    'ylim' : [910,1170],
    'Pseafloor_MPa' : 10,
    'nsteps_iceI' : 20,
    'nsteps_ocean' : 100,
    'nsteps_ref_rho' : 30,
    'nsteps_mantle' : 1500,
    'nsteps_core' : 100,
    'nsteps_colororder' : 'mcbkgrm',
    'Temps' : [245., 250., 252.5, 260., 265., 270.],
    'LineStyle' : '--',
    'wrefLine' : '--',
    'wref' : [0,34]
}

File: Ariel/PPAriel.m
function PPAriel
%PPAriel
Planet.name='Ariel';
% Planet.rho_kgm3 = 1592; %±150 Jacobson, R. A.; Campbell, J. K.; Taylor, A. H.; Synnott, S. P. (June 1992). "The masses of Uranus and its major satellites from Voyager tracking data and earth-based Uranian satellite data". The Astronomical Journal. 103 (6): 2068?2078. Bibcode:1992AJ....103.2068J. doi:10.1086/116211.
Planet.R_m = 578.9e3; %±0.6 Jacobson, R. A.; Campbell, J. K.; Taylor, A. H.; Synnott, S. P. (June 1992). "The masses of Uranus and its major satellites from Voyager tracking data and earth-based Uranian satellite data". The Astronomical Journal. 103 (6): 2068?2078. Bibcode:1992AJ....103.2068J. doi:10.1086/116211.
Planet.M_kg =1.353e21; %±0.120e21 
Planet.Tsurf_K = 60; 
Planet.Psurf_MPa = 0; 
Planet.Cmeasured = 0.306; %  Hussmann 2006 (not actually measured)
Planet.Cuncertainty = 0.08;% 
Planet.FeCore=false;
    Planet.xFeS = 0.25; %0.25
    Planet.rhoFe = 8000; %8000
    Planet.rhoFeS = 5150; %5150
% Planet.rho_sil_withcore_kgm3 = 2400; % Iess et al. 2014
Planet.rho_sil_withcore_kgm3 = 2700; 

%% &&& Orbital and plotting parameters for use in LayeredInductionResponse
Planet.peaks_Hz = [3.45599e-5 2.30405e-5 1.1519e-5];
Planet.peaks_hr = 1./Planet.peaks_Hz./3600;
Planet.f_orb = 2*pi/2.520/24/3600; % radians per second
Params.wlims = [log(0.001) log(1000)];
% Get Fourier spectrum data
try 
    load(['FTdata' Planet.name],'FTdata');
catch
    dlNow = FTdataMissingWarning();
    if dlNow; load(['FTdata' Planet.name],'FTdata'); end
end
% Planet.ionos_bounds = 100e3;
% Planet.ionosPedersen_sig = 30/100e3;
Planet.ionos_only = [];
Planet.PLOT_SIGS = true;
Planet.ADD_TRANSITION_BOUNDS = false;

% WARNING: The following line was copied from PPCallisto.m because it is
% required by PlanetProfile.m in the current version and does not appear in
% this file. An issue has been opened on GitHub. Delete this comment when
% the value has been corrected.
Planet.XH2O = 0.104; % total fraction of water in CM2; use this to compute the excess or deficit indicated by the mineralogical model


% Planet.Ocean.comp='NH3';
% % load L_IceNH3.mat
% Planet.Ocean.w_ocean_pct=5;
% Planet.Tb_K = [245 260 265]; % 5 Wt% temperatures at the bottom of the Ice Ih Van dishoek, cited by Engel et al. 1994 for Titan
% Tb = [251.5 260 270]; % 1 Wt% temperatures at the bottom of the Ice Ih
% Van dishoek, cited by Engel et al. 1994 for Titan


%%  Interior constraints imposed in Vance et al. 2014
mSi = 28.0855; mS = 32.065; mFe = 55.845; mMg = 24.305;
xOl = 0.44; % percentage of olivine - Javoy (1995) - Fe/Si = 0.909 Mg/Si = 0.531, Mg# = 0.369
%mOl = 2*((1-0.369)*58.85+0.369*24.31)+28.0855+4*16=184.295
%mPx = 2*((1-0.369)*58.85+0.369*24.31+28.0855+3*16) =244.3805
xSi = (xOl+2*(1-xOl))*mSi/(xOl*184.295+(1-xOl)*244.3805); % mass fraction of sulfur in silicates
M_Earth_kg = 5.97e24;
xSiEarth = 0.1923; % Javoy in kg/kg in Exoplanets paper20052006-xSiSolar only in mole
xK = 1; %enrichment in K
Hrad0 = 24e12*xSi/xSiEarth/M_Earth_kg;

%% Mantle Heat
%cold case  
Planet.kr_mantle = 4; % rock conductivity (Cammarano et al. 2006, Table 4)
Planet.EQUIL_Q = 0;
% Planet.Qmantle_Wm2 = 2.7e8/4/pi/Planet.R_m^2; % Chen et al. 2014
 Planet.Qmantle_Wm2 = 2.7e4/4/pi/Planet.R_m^2; % kluge
% Planet.Qmantle_Wm2 = 16e9/4/pi/Planet.R_m^2; % Howett et al. 2011
Planet.QHmantle = 0;
%hot case Qm = 2.1e11+8.5e11; %W

%% Seismic
Seismic.LOW_ICE_Q = 1; % divide Ice Q value by this number
% Seismic.mantleEOS = 'ChondriteLL_Stx11.ext'; % this was the original
% chondrite function built by svance

% this is the porous model described in the paper 2017:
Planet.POROUS_ROCK = 1;
Seismic.mantleEOS = 'echon_678_1.tab'; Planet.phi_surface = 0.8;% 
% Seismic.mantleEOS = 'pyrohp_sat_678_1.tab'; Planet.phi_surface = 0.5;% this uses the procedure implemented by F. Cammarano

% % this is the hydrated model described in the paper 2017:
% Planet.POROUS_ROCK = 0;
Seismic.mantleEOS = 'echon_hp_sat_PX678_14GPa.tab'; % this uses the procedure implemented by F. Cammarano


% Seismic.mantleEOS = 'pyrohp_sat_678_1.tab'; %  (3000) this uses the procedure implemented by F. Cammarano; this includes Ks and Gs. I had to rerun perlex (6.6.3). not sure why
% Seismic.mantleEOSname = 'pyrohpsat';

% Seismic.mantleEOS = 'CV_hhph_DEW17_nofluid_nomelt_685.tab';
% Seismic.mantleEOSname = 'CV_hhphD17nofluid';

% Seismic.mantleEOS = 'CV_hhph_DEW17_fluid_nomelt_685.tab';
% Seismic.mantleEOSname = 'CV_hhphD17fluid';

% Seismic.mantleEOS = 'CM_hhph_DEW17_nofluid_nomelt_685.tab';
% Seismic.mantleEOSname = 'CM_hhphD17nofluid';

% Seismic.mantleEOS = 'CM_hhph_DEW17_fluid_nomelt_685.tab';
% Seismic.mantleEOSname = 'CM_hhphD17fluid';

Seismic.mantleEOS = 'CI_hhph_DEW17_nofluid_nomelt_685.tab';
Seismic.mantleEOSname = 'CI_hhphD17nofluid';

% Seismic.mantleEOS = 'CI_hhph_DEW17_fluid_nomelt_685.tab';
% Seismic.mantleEOSname = 'CI_hhphD17fluid';


% Seismic.mantleEOS = 'echonhp_sat_1.tab'; % this uses the procedure implemented by F. Cammarano

Seismic.QScore = 1e4;
Seismic.SMOOTH_VROCK = 5; % smooth over N neighboring rows and columns in vp and vs



%Attenuation Parameters Based on those Described in Cammarano et al. 2006
% ice I
Seismic.B_aniso_iceI = 0.56;
Seismic.gamma_aniso_iceI = 0.2;
Seismic.g_aniso_iceI = 22; %C2006
% ice III
Seismic.B_aniso_iceIII = 0.56;
Seismic.gamma_aniso_iceIII = 0.2;
Seismic.g_aniso_iceIII = 25; 
% ice V
Seismic.B_aniso_iceV = 0.56;
Seismic.gamma_aniso_iceI = 0.2;
Seismic.g_aniso_iceV = 27; 
% ice VI
Seismic.B_aniso_iceVI = 0.56;
Seismic.gamma_aniso_iceVI = 0.2;
Seismic.g_aniso_iceVI = 28; 
% mantle
Seismic.B_aniso_mantle = 0.56;
Seismic.gamma_aniso_mantle = 0.2;
Seismic.g_aniso_mantle = 30; %C2006

%% Model Parameters
Params.savefigformat = 'epsc';
Params.foursubplots =1;
Params.HOLD = 0; % overlay previous run
Params.Legend = false;
Params.LegendPosition = 'North';
Params.ylim = [910 1170];
Params.Pseafloor_MPa = 30;
Params.nsteps_iceI = 20;
Params.nsteps_ocean = 100; 
Params.nsteps_ref_rho = 30;
Params.nsteps_mantle = 1500;
Params.nsteps_core = 100;
Params.Temps = [245 250 252.5 255 260 265 270];
Params.NOPLOTS = 0; %allows user to limit recreating plots & figures after each run


%% Run the Calculation!
% Planet.Ocean.w_ocean_pct=10;Planet.Tb_K = [272.8 272.9 273 273.1]; % pure water, temperatures at the bottom of the Ice Ih
% PlanetProfile(Planet,Seismic,Params)
% 
% 
% Params.HOLD = true;
% Params.INCLUDE_ELECTRICAL_CONDUCTIVITY = false;
% 
% Params.CALC_NEW=1;
% Params.CALC_NEW_SOUNDSPEEDS=1;
% 
% Planet.Ocean.w_ocean_pct=0; Planet.Tb_K = [273.1 273.15]; % pure water, temperatures at the bottom of the Ice Ih
% PlanetProfile(Planet,Seismic,Params)
% 
Params.INCLUDE_ELECTRICAL_CONDUCTIVITY = 1;
% 
Params.CALC_NEW =1; % Set CALC_NEW options to 0 to re-use profile data when possible. It is recommended to keep CALC_NEW=1 except when intermediate parameters such as layer thicknesses will not change between runs.
Params.CALC_NEW_REFPROFILES=1;
Params.CALC_NEW_SOUNDSPEEDS=1;
% 
Planet.Ocean.comp='MgSO4';
load L_Ice_MgSO4.mat
Planet.Ocean.fnTfreeze_K = griddedInterpolant(PPg',wwg',TT');
Params.LineStyle='--';
Params.colororder = 'cm';
Params.wrefLine = '--';
Params.wref=[0 5 10 15];
% 
Planet.Ocean.w_ocean_pct=10; Planet.Tb_K = 269.2;
outPlanet = PlanetProfile(Planet,Seismic,Params);
outWaveforms = LayeredInductionResponse(outPlanet,FTdata,Params);

Planet.Ocean.w_ocean_pct=1; Planet.Tb_K = 273.1;
outPlanet = PlanetProfile(Planet,Seismic,Params);
outWaveforms = LayeredInductionResponse(outPlanet,FTdata,Params);


% running pure water for the MgSO4 case illustrates >1oC error in the Margules parameterization
Params.LineStyle='-';
Params.colororder = 'cm';
Planet.Ocean.w_ocean_pct=0;  Planet.Tb_K =  273.1; % pure water, 
outPlanet = PlanetProfile(Planet,Seismic,Params);
% 
Planet.Ocean.comp='Seawater';
Params.LineStyle='-.';
Params.wref=[0 34 68];
Params.wrefLine = '-.';
Params.colororder = 'cm';

Planet.Ocean.w_ocean_pct=gsw_SSO; Planet.Tb_K = 270.82;
outPlanet = PlanetProfile(Planet,Seismic,Params);
outWaveforms = LayeredInductionResponse(outPlanet,FTdata,Params);

Planet.ALLOW_NEGALPHA = 0;
Planet.Ocean.w_ocean_pct=0.1*gsw_SSO; Planet.Tb_K = 272.57;
outPlanet = PlanetProfile(Planet,Seismic,Params);
outWaveforms = LayeredInductionResponse(outPlanet,FTdata,Params);


Params.HOLD = 1; % overlay previous runs
Params.LineStyle='-';
Params.colororder = 'cm';
Planet.Ocean.w_ocean_pct=0;  Planet.Tb_K =  272.74; % pure water, 
outPlanet = PlanetProfile(Planet,Seismic,Params);

Params.wrefLine =  ':';
Params.wref=[3 5 10];
Planet.Ocean.comp='NH3';
load L_IceNH3_DATA.mat
Planet.Ocean.fnTfreeze_K = griddedInterpolant(PPg',wwg',TT');

Params.CALC_NEW_REFPROFILES=1;
Params.CALC_NEW_SOUNDSPEEDS=1;
Params.HOLD = 1;
Params.LineStyle =  ':';
Planet.Ocean.w_ocean_pct=3; Planet.Tb_K = 269.535; % 0 Wt% temperatures at the bottom of the Ice Ih
%PlanetProfile(Planet,Seismic,Params)